Add support for permutting dynamic fusion group outputs to channels last format

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71173
* #71138
* __->__ #70656
* #70464
* #71144

Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

The output striding behavior is described either as contiguous or channels last, and stored on the TE Kernel.

Commit:
Add support for permutting dynamic fusion group outputs to channels last format

[ghstack-poisoned]

eellison commented on test/test_jit_fuser_te.py:
```diff
@@ -1858,6 +1858,23 @@ def eager(x, y):
         script = self.checkScript(eager, (x, y))
         self.assertAllFused(script.graph_for(x, y))
 
+    def test_channels_last_dims_dynamic(self):
```
@navahgar, @ZolotukhinM this runs into an IR Simplifier bug. I tested this on simpler functions with a single input and was able to run successfully but now hitting simplifier bug.

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"




[ghstack-poisoned]

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"




[ghstack-poisoned]

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"




[ghstack-poisoned]

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"




[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458650).

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

[ghstack-poisoned]

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458650).

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

[ghstack-poisoned]

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

[ghstack-poisoned]

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458650).

eellison commented:
With contiguous inputs
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(1, SS(-6), requires_grad=0, device=cpu)], striding_inputs_desc=[["TENSOR_CONT"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=1]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(1, 10, strides=[10, 1], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-6], striding_inputs_desc=[["TENSOR_CONT"]], striding_outputs_desc=[["TENSOR_CONT"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
with prim::TensorExprGroup_0 = graph(%x : Double(1, SS(-6), requires_grad=0, device=cpu),
      %SS_6 : int):
  %3 : Double(1, SS(-6), strides=[10, 1], requires_grad=0, device=cpu) = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
  %1 : Double(1, SS(-6), requires_grad=0, device=cpu) = aten::sigmoid(%3) # test/test_jit_fuser_te.py:2027:0
  return (%1)
```
Transposed arg and 1
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-8), SS(-8), requires_grad=0, device=cpu)], striding_inputs_desc=[["S_ONE", "S_TRAN_CONT"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(10, 10, strides=[1, 10], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-8], striding_inputs_desc=[["S_ONE", "S_TRAN_CONT"]], striding_outputs_desc=[["S_ONE", "S_TRAN_CONT"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
```
As arg input
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-9), requires_grad=0, device=cpu)], striding_inputs_desc=[["S_AS_ARG"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(10, strides=[1], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-9], striding_inputs_desc=[["S_AS_ARG"]], striding_outputs_desc=[["S_AS_ARG"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
```
Channels last 
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-11), SS(-12), SS(-13), SS(-14), requires_grad=0, device=cpu)], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = prim::Constant[value=1]()
      %20 : int = prim::Constant[value=2]()
      %21 : int = prim::Constant[value=3]()
      %22 : int = aten::__getitem__(%17, %18)
      %23 : int = aten::__getitem__(%17, %19)
      %24 : int = aten::__getitem__(%17, %20)
      %25 : int = aten::__getitem__(%17, %21)
      %9 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-14, -13, -12, -11], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]], striding_outputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]]](%x, %25, %24, %23, %22)
      -> (%9)
    block1():
      %29 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %30 : Tensor = aten::sigmoid(%29) # test/test_jit_fuser_te.py:2027:0
      %31 : (Tensor) = prim::TupleConstruct(%30)
      %28 : Tensor = prim::TupleUnpack(%31)
      -> (%28)
  return (%13)
```

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

[ghstack-poisoned]

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

[ghstack-poisoned]

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458650).

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458650).

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1111,7 +1111,82 @@ bool denseAndNonOverlapping(
+  // A contiguous Tensor of size(2, 3) with values 0-5 is layed out as:
+  // [0] [1] [2] [3] [4] [5]
+  // The same valued tensor with strides (2, 1) would be layed out like
+  // [0] [3] [1] [4] [2] [5]
```
Shouldn't the strides for this data be `(1, 2)` and not `(2, 1)`?

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1111,7 +1111,82 @@ bool denseAndNonOverlapping(
+  if (tensorOutputStrideDesc_[v->offset()] == torch::jit::StrideInput::TENSOR_CONT) {
+    return Tensor(buf, nullptr);;
+  }
+  TORCH_INTERNAL_ASSERT(tensorOutputStrideDesc_[v->offset()] == torch::jit::StrideInput::TENSOR_CONT_CHANNELS_LAST);
```
Please run clang-format

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1529,8 +1596,14 @@ void TensorExprKernel::updateOutputSizesAndStrides(
+    } else if (tensorOutputStrideDesc_[i] == torch::jit::StrideInput::TENSOR_CONT_CHANNELS_LAST) {
+      tensorOutputStrides_[i] = at::get_channels_last_strides_2d(tensorOutputSizes_[i]);
+    } else {
+      TORCH_INTERNAL_ASSERT(false);
```
Can you add an output string to this assertion?

navahgar reviewed (approved):
LGTM

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

The output striding behavior is described either as contiguous or channels last, and stored on the TE Kernel. 

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458650).

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

The output striding behavior is described either as contiguous or channels last, and stored on the TE Kernel. 

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458650).

Commit:
Update on "Add support for permutting dynamic fusion group outputs to channels last format"


Differential Revision: [D33458650](https://our.internmc.facebook.com/intern/diff/D33458650)

This follows the existing permutation logic to support strided output tensors. For now it only supports channels-last tensors. In order for this approach to work we just need to know the sorted order of strides, which allows us to support any permutation of tensor dimensions as an output in the future.

The output striding behavior is described either as contiguous or channels last, and stored on the TE Kernel. 

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458650).
